File: Data Structures & Algorithms/reverse-linked-list-ii/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {

        if (left == right)
        {
            return head;
        }
        int fp = 0;
        int sp = 0;
        ListNode* fp_head = head;
        ListNode* sp_head = head;
        ListNode* sp_prev = new ListNode(-1000,head);
        while (fp < right - 1)
        {
            if (fp >= right - left)
            {
                sp_prev = sp_head;
                sp_head = sp_head->next;
                sp += 1;
            }
            fp_head = fp_head->next;
            fp += 1;
        }
        ListNode* prev = fp_head->next;
        while (sp <= fp)
        {
            ListNode* temp = sp_head->next;
            sp_head->next = prev;
            prev = sp_head;
            sp_head = temp;
            sp += 1;
        }
        sp_prev->next = prev;

        if (left == 1)
        {
            return sp_prev->next;
        }
        else
        {
            return head;
        }
        
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        fp = head
        sp = head
        count = 0
        while fp:
            if (count >= n+1):
                sp = sp.next
            fp = fp.next
            count += 1
        
        if count == n:
            head = head.next
            return head
        
        temp = sp.next
        sp.next = temp.next

        return head
        


        